File: algorithms/647. Palindromic Substrings/solution.py
class Solution(object):
    def countSubstrings(self, s):
        """
        :type s: str
        :rtype: int
        """
        n = len(s)
        dp = [[0] * n for i in xrange(n)]
        count = 0
        for end in xrange(n):
            dp[end][end] = 1
            count += 1
            for start in xrange(end):
                if s[start] == s[end] and (start+1 >= end-1 or dp[start+1][end-1]):
                    count += 1
                    dp[start][end] = 1
        return count

# 130 / 130 test cases passed.
# Status: Accepted
# Runtime: 512 ms

File: algorithms/647. Palindromic Substrings/solution.cc
class Solution {
public:
    int countSubstrings(string s) {
        // @ hao-cai
        int cnt = 0, n = s.size();
        for ( int i = 0; i < n; ++i ) {
            for ( int j = 0; i-j >= 0 && i+j < n && s[i-j] == s[i+j]; ++j )
                ++cnt;
            for ( int j = 0; i-j-1 >= 0 && i+j < n && s[i-j-1] == s[i+j]; ++j )
                ++cnt;
        }
        return cnt;
    }
};

// 130 / 130 test cases passed.
// Status: Accepted
// Runtime: 6 ms
// beats 67.19 %
